### PR TITLE
feat: Update pydantic to v2

### PR DESCRIPTION
**Pull request by kcze**
### Background

Agent Server is incompatible with AutoGPT&Forge because they use pydantic v2 and v1 respectively.

### Changes 🏗️

Update Pydantic dependency of `autogpt`, `forge` and `benchmark` to `^2.7`
[Pydantic Migration Guide](https://docs.pydantic.dev/2.7/migration/)

- Rename and update deprecated functions
- Default values are strict and aren't inferred; `None` has to be specified explicitly for `Optional` and model fields without defaults are required (Ellipsis `...` for required fields is deprecated)
- `Config` class for specifying model configuration is deprecated, now `model_config` class attribute needs to be set to `ConfigDict` object with configuration
- Custom `kwargs` (`extra`) can no longer be passed to `Field`
- Forked `agent-protocol` and updated *only* client package: https://github.com/kcze/agent-protocol
- Removed `ModelContainer` in `BaseAgent`, component configuration is now serialized using Pydantic v2 functions directly

### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

### CODE REVIEW

File: forge/forge/models/config.py
Comment (kcze) [inline forge/forge/models/config.py]: There's no more `extra` and custom values cannot be passed directly to `Field`, so I used `metadata` instead; https://github.com/pydantic/pydantic/discussions/2419#discussioncomment-7228409

File: autogpt/autogpt/agents/prompt_strategies/one_shot.py
Comment (kcze) [inline autogpt/autogpt/agents/prompt_strategies/one_shot.py:32]: Ellipsis `...` is no longer needed, any field without default value/factory is required.

File: benchmark/pyproject.toml
Comment (kcze) [inline benchmark/pyproject.toml:21]: `multidict 6.0.4` fails to install and is a dependency of some package, this has been a reoccurring issue.

File: autogpt/autogpt/app/config.py
Comment (Pwuts) [inline autogpt/autogpt/app/config.py]: According to the docs, `@field_validator` implies `@classmethod`

File: forge/pyproject.toml
Comment (Pwuts) [inline forge/pyproject.toml]: `agent-protocol-client` is not used in `forge`

File: forge/forge/models/config.py
Comment (Pwuts) [inline forge/forge/models/config.py]: ? in what scenario is validation not desirable?

File: autogpt/autogpt/app/config.py
Comment (Pwuts) [inline autogpt/autogpt/app/config.py:96]: Why remove the `if` statement?

File: benchmark/agbenchmark/app.py
Comment (Pwuts) [inline benchmark/agbenchmark/app.py]: `challenge_spec_file` is already a `Path`
```suggestion
            challenge_spec_file.read_text()
```

File: benchmark/agbenchmark/challenges/base.py
Comment (Pwuts) [inline benchmark/agbenchmark/challenges/base.py]: ?

File: benchmark/agbenchmark/challenges/builtin.py
Comment (Pwuts) [inline benchmark/agbenchmark/challenges/builtin.py]: ?

File: benchmark/agbenchmark/challenges/builtin.py
Comment (Pwuts) [inline benchmark/agbenchmark/challenges/builtin.py]: `spec_file` is already a `Path`:
```suggestion
        challenge_spec = BuiltinChallengeSpec.model_validate_json(spec_file.read_text())
```

File: benchmark/pyproject.toml
Comment (Pwuts) [inline benchmark/pyproject.toml]: 1. This is in the wrong place in the list
2. That fork should be in the Significant-Gravitas org

File: benchmark/pyproject.toml
Comment (Pwuts) [reply benchmark/pyproject.toml:21]: Please add this as a comment. Be sure to mention *which* package has `multidict` as a dependency, so we know to remove it if we remove the parent dependency.

File: benchmark/reports/format.py
Comment (Pwuts) [inline benchmark/reports/format.py]: `report_json_file` should already be a `Path`; I think it is annotated incorrectly. You can fix the parameter annotation instead.
```suggestion
    report = Report.model_validate_json(report_json_file.read_text())
```

File: forge/forge/agent/base.py
Comment (Pwuts) [inline forge/forge/agent/base.py]: According to the docs, `@field_validator` implies `@classmethod`:
```suggestion
    @field_validator("use_functions_api")
```

File: forge/forge/components/code_executor/code_executor.py
Comment (Pwuts) [inline forge/forge/components/code_executor/code_executor.py:88]: Nit: for aesthetics I would format it like this
```suggestion
        default_container_name = (
            self.config.model_fields["docker_container_name"].default
        )
```

File: forge/forge/llm/providers/anthropic.py
Comment (Pwuts) [inline forge/forge/llm/providers/anthropic.py]: The `# type: ignore` is necessary because the type checker can't see through the `model_dump` call: https://github.com/Significant-Gravitas/AutoGPT/actions/runs/9752798470/job/26916967911?pr=7280#step:7:26

File: forge/forge/llm/providers/schema.py
Comment (Pwuts) [inline forge/forge/llm/providers/schema.py:207]: On `strict` mode, the type checker complains that "ignore" is not a valid value:
```suggestion
    model_config = ConfigDict(extra=Extra.ignore)
```

File: forge/forge/models/config.py
Comment (Pwuts) [inline forge/forge/models/config.py]:
```suggestion
def _get_field_metadata(field: FieldInfo, key: str, default: Any = None) -> Any:
```

File: forge/forge/models/config.py
Comment (Pwuts) [inline forge/forge/models/config.py]: you *could* consider:
```suggestion
    field_info.metadata.append(("user_configurable", True))
```
this would allow `dict(field_info.metadata).get("user_configurable", False)`

File: forge/forge/models/config.py
Comment (Pwuts) [inline forge/forge/models/config.py]:
```suggestion
```

File: forge/forge/models/config.py
Comment (Pwuts) [inline forge/forge/models/config.py]: 🤔

**Review by Pwuts — CHANGES_REQUESTED** [review]
Legendary effort. Thank you very much for pulling this.

Question marks and suggestions: ⬇️ (or ⬆️ in the app)

File: forge/forge/llm/providers/openai.py
Comment (Pwuts) [inline forge/forge/llm/providers/openai.py]: This should not be optional

**Review by Pwuts — CHANGES_REQUESTED** [review]

File: autogpt/autogpt/app/config.py
Comment (kcze) [reply autogpt/autogpt/app/config.py]: removed

File: forge/pyproject.toml
Comment (kcze) [reply forge/pyproject.toml]: removed

File: forge/forge/models/config.py
Comment (kcze) [reply forge/forge/models/config.py]: Removed; I was trying to solve some problem skipping validation in some cases but it works now when always validating.

File: autogpt/autogpt/app/config.py
Comment (kcze) [reply autogpt/autogpt/app/config.py:96]: 🤷 reverted

File: benchmark/agbenchmark/challenges/base.py
Comment (kcze) [reply benchmark/agbenchmark/challenges/base.py]: Reverted, since pydantic v2 requires values without default I think this was causing problems at some point.

File: benchmark/agbenchmark/challenges/builtin.py
Comment (kcze) [reply benchmark/agbenchmark/challenges/builtin.py]: reverted

File: benchmark/pyproject.toml
Comment (kcze) [reply benchmark/pyproject.toml]: transferred ownership & reordered

File: forge/forge/llm/providers/schema.py
Comment (kcze) [reply forge/forge/llm/providers/schema.py:207]: Doesn't complain for me, and when I change it to `Extra.*`:

> Extra is deprecated. Use literal values instead (e.g. `extra='allow'`)

File: forge/forge/models/config.py
Comment (kcze) [reply forge/forge/models/config.py]: ~~I changed to this and simplified `_get_field_metadata`~~
edit: not doing this, it can lead errors if `metadata` contains values incompatible with `dict()`

File: forge/forge/models/config.py
Comment (kcze) [reply forge/forge/models/config.py]: 🤦 I meant `instance[name]` but that isn't possible. Changed to:
```py
for name, value in instance.__dict__.items():
        field = instance.model_fields[name]
```

File: forge/forge/llm/providers/openai.py
Comment (kcze) [reply forge/forge/llm/providers/openai.py]: reverted

File: benchmark/pyproject.toml
Comment (Pwuts) [inline benchmark/pyproject.toml:12]: - [ ] https://github.com/Significant-Gravitas/agent-protocol/pull/2
- [ ] Make `v1.2.0` tag after merging above PR
- [ ] Pin this dependency to `rev = "v1.2.0"` to prevent it from auto-updating to dev versions

Comment (Pwuts) [reply benchmark/pyproject.toml:12]: This *can* wait until after this PR is merged

**Review by Pwuts — APPROVED** [review]
LGTM!

**Pull request merged**
